PaletteEdit: add changelog

ramonjd opened the pull request:
## What?

This PR  adds a changelog for alterations merged in #45681

It also add a test case for PaletteEdit to ensure that the correct colorpicker is displayed. Picker functionality is covered in existing tests, e.g., packages/components/src/color-palette/test/index.tsx

Update: moving JS unit tests for gradient to a follow up since there is a `"not wrapped in act(...)" warning` that results in a fail. Probably related to the Popover. Tried many available async approaches. Maybe rejigging the component, mocking Popover might work.

ramonjd committed:
Initial commit:
- Add changelog for changes merged in #45681
- Test case for Popover

ramonjd committed:
Change log entry

ramonjd committed:
Formatting

andrewserong commented on packages/components/CHANGELOG.md:
```diff
@@ -12,10 +12,11 @@
--   `InputControl`, `NumberControl`, `UnitControl`: Add `help` prop for additional description ([#45931](https://github.com/WordPress/gutenberg/pull/45931)).
--   `BorderControl`, `ColorPicker` & `QueryControls`: Replace bottom margin overrides with `__nextHasNoMarginBottom` ([#45985](https://github.com/WordPress/gutenberg/pull/45985)).
--   `CustomSelectControl`, `UnitControl`: Add `onFocus` and `onBlur` props ([#46096](https://github.com/WordPress/gutenberg/pull/46096)).
+- `TabPanel`: Add ability to set icon only tab buttons ([#45005](https://github.com/WordPress/gutenberg/pull/45005)).
+- `InputControl`, `NumberControl`, `UnitControl`: Add `help` prop for additional description ([#45931](https://github.com/WordPress/gutenberg/pull/45931)).
+- `BorderControl`, `ColorPicker` & `QueryControls`: Replace bottom margin overrides with `__nextHasNoMarginBottom` ([#45985](https://github.com/WordPress/gutenberg/pull/45985)).
+- `CustomSelectControl`, `UnitControl`: Add `onFocus` and `onBlur` props ([#46096](https://github.com/WordPress/gutenberg/pull/46096)).
+- `PaletteEdit`: Global styles: add onChange actions to color palette items [#45681](https://github.com/WordPress/gutenberg/pull/45681)
```
Tiniest of tiny nits:

* There's a whitespace change between the `-` and the name of the component. Was that intentional? It looks different to the other entries elsewhere in the changelog.
* Missing additional parentheses around the PR, and a full-stop at the end.

ramonjd replied:
Good spotting. That was my IDE getting assertive. Changing....

andrewserong commented on packages/components/CHANGELOG.md:
```diff
@@ -27,6 +28,8 @@
 -   `LinkedButton`: remove unnecessary `span` tag ([#46063](https://github.com/WordPress/gutenberg/pull/46063))
 -   NumberControl: refactor styles/tests/stories to TypeScript, replace fireEvent with user-event ([#45990](https://github.com/WordPress/gutenberg/pull/45990)).
 -   `useBaseField`: Convert to TypeScript ([#45712](https://github.com/WordPress/gutenberg/pull/45712)).
+-   `PaletteEdit`: add follow up changelog for #45681 and tests [#46095](https://github.com/WordPress/gutenberg/pull/46095)
```
Tiny nit: Missing parentheses around the PR link and a full-stop.

andrewserong commented on packages/components/src/palette-edit/index.js:
```diff
@@ -101,6 +101,11 @@ function ColorPickerPopover( {
 			offset={ 20 }
 			className="components-palette-edit__popover"
 			onClose={ onClose }
+			aria-label={
+				! isGradient
+					? __( 'Select a new color' )
+					: __( 'Select a new gradient' )
+			}
```
I don't know too much about `aria-label` usage, but from a quick look at [the MDN docs](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-label), it sounds like it's intended to be used only on interactive elements (to indicate how an element can be interacted with) rather than to wrappers?

I was wondering if the addition here is so that the tests have something to hook onto for determining whether or not `ColorPicker` or `CustomGradientPicker` is rendered. If so, could it would work to look up the label "Color" for the color picker, or the label (or part of the label) `Gradient control point at position" for the custom gradient picker?

Markup for color picker:

![image](https://user-images.githubusercontent.com/14988353/205197327-22f63404-710e-4471-8122-b65ed51707c7.png)

Markup for gradient picker:

<img width="649" alt="image" src="https://user-images.githubusercontent.com/14988353/205197246-25ab239d-cb24-4aed-afaf-55b053f9264a.png">

ramonjd replied:
You're right. Excellent point. Thanks for making this PR better @andrewserong !

andrewserong reviewed (commented):
Thanks for the follow-up PR! Just left a couple of nits, and a question about whether we can use the existing labels within each component as targets for the test, instead of adding an additional `aria-label` value at the wrapper level, since it sounds like `aria-label` is intended to be used on interactive elements rather than wrappers, as far as I can tell.

Let me know if I'm way off, though!

ramonjd committed:
Changelog formatting
Relying on existing control aria labels to test for the presence of gradient and color pickers rather than the container.

ramonjd committed:
Export no longer required

andrewserong commented on packages/components/src/palette-edit/test/index.js:
```diff
@@ -61,3 +66,46 @@ describe( 'getNameForPosition', () => {
+		slugPrefix: '',
+	};
+
+	it( 'opens color selector for color palettes', async () => {
```
Thanks for updating so quickly!

The tests are passing for me locally, but I noticed the Github Action appears to be failing. I kicked it off again, but if it still fails, would it be worth trying to remove the `async` here? I wasn't sure if it's needed since the other test doesn't have it 🤔

ramonjd replied:
That was an oversight from previous testing. Thanks again for 👀

andrewserong replied:
Hrm, for some reason the test is still complaining 🤔
Happy to help debug on Monday!

ramonjd replied:
Yeah it's curious. `fireEvent` is already wrapped in `act()` apparently. 
Tried using async `getBy*` queries as well. Still failed. 🤷 
Oh well.

ramonjd committed:
Remove unnecessary async function dec

ramonjd committed:
Testing async queries using findByLabelText because the CI tests are failing (not local)

ramonjd committed:
Let's try this.

ramonjd committed:
Let's try this.

ramonjd committed:
What about this?

ramonjd committed:
Why not

ramonjd commented:
If this round of testing doesn't work I think we should separate the requirement for a changelog for https://github.com/WordPress/gutenberg/pull/45681 and the tests.

The fact that it only fails on CI is troubling. 😄

andrewserong commented:
> If this round of testing doesn't work I think we should separate the requirement for a changelog for https://github.com/WordPress/gutenberg/pull/45681 and the tests.

Good idea, thanks for persevering!

ramonjd committed:
And so it continues

ramonjd commented:
It appears to me to be related to the Popover component, which is a child of the PaletteEdit component. 

The Popover component's tests require the use of [async testing code](https://github.com/WordPress/gutenberg/blob/705ab44f2c2797d75740533a85030c194b4fb040/packages/components/src/popover/test/index.tsx#L96), but we're not taking that into account here. 

The only thing I can think of is to mock the component somehow 🤷

andrewserong commented:
> The Popover component's tests require the use of [async testing code](https://github.com/WordPress/gutenberg/blob/705ab44f2c2797d75740533a85030c194b4fb040/packages/components/src/popover/test/index.tsx#L96), but we're not taking that into account here.

Oh, good catch. Would it work to use that wrapper (`renderAsyc`) in this test, too, or do you think it has to be at the `Popover` component level, rather than at the parent component?

ramonjd committed:
Removing tests so we can debug them separately

ramonjd commented:
> Would it work to use that wrapper (renderAsyc) in this test, too, or do you think it has to be at the Popover component level, rather than at the parent component?

Tried it 😄 I think the latter since we're changing props sent to the Popover, which we're then querying. 

Strange that the color test passed, but not the gradient 🤷 

I'll look into it more deeply in a follow up. 

Thanks for reviewing again.

andrewserong reviewed (approved):
Just an approval for the changelog version 😄

ramonjd committed:
Reinstate test for color only. Gradient needs some async magic

ramonjd commented:
I readded the basic test to get some coverage since it was passing fine. 
Only opening the gradient panel fails, I assume, because it relies on a prop change to switch the popover content. I don't know.

andrewserong commented:
> I readded the basic test to get some coverage since it was passing fine.

Good thinking 👍
